### DISCUSSION TITLE
How to use Etsy API v3?

### DISCUSSION THREAD

**Discussion by strong-728** [etsy/open-api#1297] [Q&A]
I want to get the order items using etsy api v3, but I am not sure how I can get the order items.
Now I can see only the keystring and shared secret on my etsy developer account.
Could you let me know the way that I can get the orders items of my Etsy account by using api?
In this screenshot, the data I want to get on Etsy api v3 is same data by clicking download csv button.

![image](https://github.com/user-attachments/assets/c17cb474-3ddb-4330-b072-eb41d1da883c)

**Comment by Clarry100** [comment]
This is exactly what I am trying to achieve at the moment. 
But I've hit a roadblock with getting oauth access token back from etsy servers.

The endpoint required is:
https://openapi.etsy.com/v3/application/shops/{shop_id}/receipts
Requires: oauth2 + transactions_r
I have my shop_id

Hoping to find some guidance on this in this thread.

**Reply by strong-728** [reply]
Hello Clarry, Thanks for your reply.
I want to display the data of orders items to my admin dashboard after save to the db.
So faced challenge is to save data into the db the order items by using open api v3.
But I am not sure how can I get the data from the api. This is my problem. :(

**Reply by Clarry100** [reply]
I am using Python code developed using Visual Studio to build my API applet. Once I get it working I will compile the python code into .EXE code that can run under Windows 11. I then intend to invoke that applet from my MSAccess DB so all i need is a single button click in Access to download the orders data in CSV format.

The first thing you need is to have an API app approved by etsy. When that is done you will receive an API keystring and a shared secret code for your app. You need to have both of these in place to proceed any further. If you already have this then you are good to go further.

Once you have those then you will need oauth access token and transactions_r scope in place.

This is where it gets deeper and difficult.
You also need to have a domain setup somewhere to receive data communication from etsy servers for them to pass back your oauth codes. I am trying to use Google Cloud and my own domain for this purpose.

One your Domain AND Request callback URI is approved by etsy. You can then make the required API calls to firstly retrieve your oauth keystring and then use that oauth keystring in the API call to the getShopReceipts endpoint I gave above.

If your subsequent call to the getShopReceipts endpoint is successful the servers will return the data you have requested in JSON format.
You then need to parse that JSON data into a format that is readable by your database. In my case I aim to use CSV format which I can already upload into my MSAccess database.

I have reached the point where I am trying to get etsy to approve my callback URI. So far no luck. But I cannot proceed further until I have that in place.

I am hoping that an experienced etsy API developer will see our help requests here and be able to give both of us some pointers on a way ahead.

**Reply by Clarry100** [reply]
![Etsy_Orders](https://github.com/user-attachments/assets/52da91db-6a6b-4461-86e1-c5f4578dd0c6)
Here is a screenshot of my little control panel in my Access DB. I can just select the required input parameters (or leave them at defaults) and then click the "Fetch Etsy Orders Data" button. This will run my API applet and retrieve the CSV data ready to upload in Access.
At least that's the plan!!

**Comment by strong-728** [comment]
Thanks for your reply.
I need the JS example code to implement the api for getting orders items data when I assume I already have the permission of the Etsy api.
Please help me if you are Etsy api v3 expert.

**Reply by Clarry100** [reply]
Have you submitted an application to etsy for an API app?
If so, and they have approved it, there will be an entry on your developers console showing the approved app and also showing your API keystring and your shared secret.
This is sufficient for you to use those endpoints that do not require oauth access keys. So you can develop APIs to retrieve public info about listings etc. eg all active listings.
However, for the API you wish to develop to retrieve etsy order details this gathers private info about your buyer so it requires the oauth permissions and access keys. Also the "transactions_r" scope is required. These keys are not issued by etsy as a result of an application from you. You have to code up an oauth flow which, if successful, results in them providing an oauth access keystring. You can then make use of this key along with your already existing API key to retrieve the data you are looking for. That data is returned in JSON format, so you then have to translate it into a format usable by your database. I am using CSV format.
This is what I am struggling to achieve at the present. So far I have reached the point where I have managed to get etsy to approve my domain and my URI which is used by etsy to pass back the oauth data required. I have not yet reached the point when I can receive the oauth keystring.
So I am by no means an expert in this stuff as I am struggling myself.
Also I am using Python and the Visual Studio development environment to do my coding. I am not familiar with JavaScript language.
The code is quite complex and mine is by no means ready to share as its not working yet.
There is much more to it than just the code. You need to set up what equates to a public webserver which allows etsy to interact with your code to pass back the requested oauth data. This in itself is a substantial task.
You need to study and understand the developer API reference manual.
https://developer.etsy.com/documentation/reference/#operation/getShopReceipts

**Comment by Wilderbill** [comment]
See:  https://github.com/slider1234/ETSY-API-Auto-Authenticator
